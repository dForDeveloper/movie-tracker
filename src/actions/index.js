export const addMovie = (movie) => ({
  type: 'ADD_MOVIE',
  movie
});

export const setUser = (user) => ({
  type: 'SET_USER',
  user
});

export const logoutUser = () => ({
  type: 'LOGOUT_USER'
});

export const getFavorites = (favorites) => ({
  type: 'GET_FAVORITES',
  favorites
});

export const addFavorite = (movie) => ({
  type: 'ADD_FAVORITE',
  movie
});