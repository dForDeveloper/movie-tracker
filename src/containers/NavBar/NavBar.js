import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { setUser, setFavorites } from '../../actions';

export class NavBar extends Component {
  handleClick = () => {
    this.props.setUser({});
    this.props.setFavorites([]);
    localStorage.removeItem('user');
    localStorage.removeItem('favorites');
  }

  getUserLinks = () => {
    const { currentUser, location } = this.props
    if (currentUser.name) {
      return (
        <div className="user-links">
          <p className="greeting">Hello,<span className="username">
           { currentUser.name }
          </span></p>
          <span className="nav-left">
          {location.pathname !== '/' && <Link to='/' id="item-left">Home</Link>}
          {location.pathname === '/' && <Link to='/favorites' id="item-left">View Favorites</Link>}
          </span>
          <span className="nav-right">
            <Link to='/' id="logout-link" onClick={this.handleClick}>Log Out</Link>
          </span>
        </div>
      );
    } else {
      return (
        <div className="user-links">
          <span className="nav-left">
            {location.pathname !== '/' && <Link to='/' id="item-left">Home</Link>}
            <Link to='/sign-up' id="item-left">Sign Up</Link>
            <Link to='/login' id="item-right">Log In</Link>
          </span>
          <span className="nav-right">
            <p>Log in to save your favorite movies.</p>
          </span>
        </div>
      );
    }
  }

  render() {
    return (
      <nav className="NavBar">
        {this.getUserLinks()}
      </nav>
    );
  }
}

export const mapStateToProps = (state) => ({
  currentUser: state.currentUser
});

export const mapDispatchToProps = (dispatch) => ({
  setUser: (user) => dispatch(setUser(user)),
  setFavorites: (favorites) => dispatch(setFavorites(favorites))
}); 

export default connect(mapStateToProps, mapDispatchToProps)(NavBar)

NavBar.propTypes = {
  currentUser: PropTypes.object,
  setFavorites: PropTypes.func,
  setUser: PropTypes.func,
};