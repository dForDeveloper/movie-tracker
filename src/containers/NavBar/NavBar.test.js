import React from 'react';
import { shallow } from 'enzyme';
import { NavBar, mapStateToProps, mapDispatchToProps } from './NavBar';
import { setUser, setFavorites } from '../../actions';

const setUserMock = jest.fn();
const setFavoritesMock = jest.fn();
let mockCurrentUser = { name: 'shannon' }

describe('NavBar', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <NavBar 
        setUser={setUserMock} 
        setFavorites={setFavoritesMock} 
        currentUser={mockCurrentUser} 
      />
    );
  });

  describe('NavBar container', () => {
    it('should match the snapshot when the user is not logged in', () => {
      expect(wrapper).toMatchSnapshot();
    });
    
    it('should call setUser and setFavorites when handleClick is invoked', () => {
      wrapper.find('.logout-link').simulate('click');
      expect(setUserMock).toBeCalled();
      expect(setFavoritesMock).toBeCalled();
    });

    it('should match the snapshot when the user is not logged in', () => {
      mockCurrentUser = {};
      wrapper = shallow(
        <NavBar 
          setUser={setUserMock} 
          setFavorites={setFavoritesMock} 
          currentUser={mockCurrentUser} 
        />
      );
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('mapStateToProps', () => {
    it(`should return an object with a currentUser object`, () => {
      const mockState = {
        movies: [{ id: 234567, title: 'A Star is Born' }],
        currentUser: { id: 1, name: 'Jeo' },
        favorites: [],
        showLoginPrompt: false
      }
      const expected = { currentUser: { id: 1, name: 'Jeo' } };
      const result = mapStateToProps(mockState);
      expect(result).toEqual(expected);
    });    
  });

  describe('mapDispatchToProps', () => {
    let dispatchMock = jest.fn()
    const result = mapDispatchToProps(dispatchMock);

    it('should call dispatch when setUser is called', () => {
      const expected = setUser({ id: 1, name: 'Jeo' });
      result.setUser({ id: 1, name: 'Jeo' });
      expect(dispatchMock).toHaveBeenCalledWith(expected);
    });

    it('should call dispatch when setFavorites is called', () => {
      const expected = setFavorites([123456, 234567]);
      result.setFavorites([123456, 234567]);
      expect(dispatchMock).toHaveBeenCalledWith(expected);
    });
  });
});