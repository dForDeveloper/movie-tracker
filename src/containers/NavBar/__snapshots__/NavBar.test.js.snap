// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar NavBar container should match the snapshot when the user is not logged in 1`] = `
<nav
  className="NavBar"
>
  <div
    className="user-links"
  >
    <p
      className="greeting"
    >
      Hello,
      <span
        className="username"
      >
        shannon
      </span>
    </p>
    <span
      className="nav-left"
    >
      <Link
        id="item-left"
        replace={false}
        to="/favorites"
      >
        View Favorites
      </Link>
    </span>
    <span
      className="nav-right"
    >
      <Link
        id="logout-link"
        onClick={[Function]}
        replace={false}
        to="/"
      >
        Log Out
      </Link>
    </span>
  </div>
</nav>
`;

exports[`NavBar NavBar container should match the snapshot when the user is not logged in 2`] = `
<nav
  className="NavBar"
>
  <div
    className="user-links"
  >
    <span
      className="nav-left"
    >
      <Link
        id="item-left"
        replace={false}
        to="/"
      >
        Home
      </Link>
      <Link
        id="item-left"
        replace={false}
        to="/sign-up"
      >
        Sign Up
      </Link>
      <Link
        id="item-right"
        replace={false}
        to="/login"
      >
        Log In
      </Link>
    </span>
    <span
      className="nav-right"
    >
      <p>
        Log in to save your favorite movies.
      </p>
    </span>
  </div>
</nav>
`;
